agent/agent.go

```diff
@@ -398,15 +408,42 @@ func (a *Agent) newExecutionProviderParams(req *agentapi.DeployRequest, tmpFile
 	return params, nil
 }
 
+func (a *Agent) setupSignalHandlers() {
+	go func() {
+		signal.Reset(syscall.SIGINT, syscall.SIGTERM, syscall.SIGUSR1, syscall.SIGUSR2, syscall.SIGHUP)
+		c := make(chan os.Signal, 1)
+		signal.Notify(c, os.Interrupt, syscall.SIGTERM, syscall.SIGINT, syscall.SIGQUIT)
+
+		for {
+			switch s := <-c; {
+			case s == syscall.SIGTERM || s == os.Interrupt || s == syscall.SIGQUIT:
+				fmt.Fprintf(os.Stdout, "Caught signal [%s], requesting clean shutdown", s.String())
+				a.shutdown()
+				os.Exit(0)
+
+			default:
+				os.Exit(0)
+			}
+		}
+	}()
+}
+
 func (a *Agent) shutdown() {
 	if atomic.AddUint32(&a.closing, 1) == 1 {
+		if a.provider != nil {
+			err := a.provider.Undeploy()
```

Comment:
This is going to probably result in `Undeploy()` being called twice when a workload has been deployed, which would require an additional mutex to ensure that call is idempotent.

Reply:
I can't see how this will call undeploy twice.

Reply:
I believe that in at least some cases it will get called once when `agentint.<id>.undeploy` is handled and a second time when the agent is shutting down.

Reply:
fixed